File: backend/app/three_mf.py
"""读取 3MF，并按原始坐标检查尺寸。"""
from pathlib import PurePosixPath
from io import BytesIO
from zipfile import BadZipFile, ZipFile
import xml.etree.ElementTree as ET

import lib3mf
import numpy as np
from fastapi import HTTPException

from .stl import assert_dimensions, parse_slots

CORE = "{http://schemas.microsoft.com/3dmanufacturing/core/2015/02}"
# 这些上限和前端一致，防止小压缩包解开后占满内存。
MAX_EXPANDED = 128 * 1024 * 1024
MAX_XML = 32 * 1024 * 1024
MAX_ENTRIES = 2048
MAX_INSTANCES = 4096
MAX_TRIANGLES = 500_000
MAX_VERTICES = 1_500_000
UNITS = {"micron": .001, "millimeter": 1, "centimeter": 10, "inch": 25.4, "foot": 304.8, "meter": 1000}
REVIEWABLE_PARSE_REASONS = (
    "expanded archive too large",
    "XML too large",
    "expanded mesh too large",
    "assembly too complex",
    "beam lattice geometry unsupported",
    "mixed part units unsupported",
    "mesh geometry required",
    "incomplete components",
)


def unsafe_zip_entry(item):
    name = item.filename
    return (name.startswith("/") or "\\" in name or ":" in name
            or ".." in PurePosixPath(name).parts or item.flag_bits & 1
            or item.compress_type not in (0, 8))


def local_name(tag):
    return tag.rsplit("}", 1)[-1].lower()


def read_parts(path):
    parts, names, total, xml_size = {}, set(), 0, 0
    with ZipFile(path) as package:
        entries = package.infolist()
        if len(entries) > MAX_ENTRIES:
            raise ValueError("too many entries")
        for item in entries:
            name = item.filename
            if name in names or unsafe_zip_entry(item):
                raise ValueError("unsafe package path")
            names.add(name)
            total += item.file_size
            if total > MAX_EXPANDED or item.compress_type not in (0, 8):
                raise ValueError("expanded archive too large or unsupported compression")
            is_xml = name.lower().endswith((".model", ".xml", ".rels"))
            if is_xml:
                xml_size += item.file_size
                if xml_size > MAX_XML:
                    raise ValueError("XML too large")
            # 在内存里逐个检查，不把包内文件落到磁盘。
            with package.open(item) as source:
                content = source.read(item.file_size + 1)
                if len(content) != item.file_size:
                    raise ValueError("invalid expanded length")
            if not is_xml:
                continue
            if b"<!DOCTYPE" in content.upper() or b"<!ENTITY" in content.upper() or b"\x00" in content:
                raise ValueError("DTD and non UTF-8 XML unsupported")
            content.decode("utf-8-sig")
            document = ET.iterparse(BytesIO(content), events=("start", "end"))
            depth = 0
            for event, _node in document:
                depth += 1 if event == "start" else -1
                if depth > 100:
                    raise ValueError("XML nesting too deep")
            root = document.root
            if name.lower().endswith(".rels") and any(e.get("TargetMode", "").lower() == "external" for e in root):
                raise ValueError("external relationships unsupported")
            if name.lower().endswith(".model"):
                if root.tag != CORE + "model" or root.get("unit", "millimeter") not in UNITS:
                    raise ValueError("invalid model root or unit")
                # 晶格可能伸到网格外，先交给人工确认。
                if any(node.tag.rsplit("}", 1)[-1].lower() == "beamlattice" for node in root.iter()):
                    raise ValueError("beam lattice geometry unsupported")
                parts["/" + name] = root
    return parts


def transform(text):
    result = np.eye(4)
    if text is not None:
        values = np.array([float(value) for value in text.split()])
        if values.size != 12 or not np.isfinite(values).all():
            raise ValueError("invalid transform")
        result[:, :3] = values.reshape(4, 3)
    return result


def inspect_3mf(path):
    with parse_slots:
        try:
            parts = read_parts(path)
            wrapper = lib3mf.get_wrapper()
            model = wrapper.CreateModel()
            reader = model.QueryReader("3mf")
            # SDK 负责格式和零件引用，包里的切片参数不用。
            reader.ReadFromFile(str(path))
            root = parts[model.RootModelPart().GetPath()]
            unit = UNITS[root.get("unit", "millimeter")]
            raw_objects = {}
            for part, document in parts.items():
                if UNITS[document.get("unit", "millimeter")] != unit:
                    raise ValueError("mixed part units unsupported")
                for obj in document.findall(f"{CORE}resources/{CORE}object"):
                    key = (part, int(obj.attrib["id"]))
                    if key in raw_objects:
                        raise ValueError("duplicate object")
                    raw_objects[key] = obj
            minimum, maximum = np.full(3, np.inf), np.full(3, -np.inf)
            instances = triangles = vertex_count = 0
            meshes = {}

            def visit(obj, matrix, ancestors):
                nonlocal instances, triangles, vertex_count, minimum, maximum
                instances += 1
                key = (obj.PackagePart().GetPath(), obj.GetModelResourceID())
                if instances > MAX_INSTANCES or len(ancestors) >= 32:
                    raise ValueError("assembly too complex")
                if key in ancestors:
                    raise ValueError("cyclic assembly unsupported")
                raw = raw_objects[key]
                if obj.IsMeshObject():
                    mesh = model.GetMeshObjectByID(obj.GetUniqueResourceID())
                    triangles += mesh.GetTriangleCount()
                    vertex_count += mesh.GetVertexCount()
                    if triangles > MAX_TRIANGLES or vertex_count > MAX_VERTICES:
                        raise ValueError("expanded mesh too large")
                    if key not in meshes:
                        vertices = np.array([[float(v.attrib[axis]) for axis in ("x", "y", "z")]
                                             for v in raw.findall(f"{CORE}mesh/{CORE}vertices/{CORE}vertex")])
                        faces = np.array([[int(t.attrib[axis]) for axis in ("v1", "v2", "v3")]
                                          for t in raw.findall(f"{CORE}mesh/{CORE}triangles/{CORE}triangle")])
                        if (len(vertices) != mesh.GetVertexCount() or len(faces) != mesh.GetTriangleCount()
                                or len(vertices) < 3 or len(faces) == 0 or not np.isfinite(vertices).all()
                                or np.min(faces) < 0 or np.max(faces) >= len(vertices)):
                            raise ValueError("invalid mesh")
                        meshes[key] = vertices
                    points = (meshes[key] @ matrix[:3, :3] + matrix[3, :3]) * unit
                    if not np.isfinite(points).all():
                        raise ValueError("invalid transformed vertices")
                    minimum = np.minimum(minimum, points.min(axis=0))
                    maximum = np.maximum(maximum, points.max(axis=0))
                elif obj.IsComponentsObject():
                    components = model.GetComponentsObjectByID(obj.GetUniqueResourceID())
                    raw_components = raw.findall(f"{CORE}components/{CORE}component")
                    if len(raw_components) != components.GetComponentCount():
                        raise ValueError("incomplete components")
                    for index, component in enumerate(raw_components):
                        child = components.GetComponent(index).GetObjectResource()
                        visit(child, transform(component.get("transform")) @ matrix, ancestors | {key})
                else:
                    raise ValueError("mesh geometry required")

            items = model.GetBuildItems()
            raw_items = root.findall(f"{CORE}build/{CORE}item")
            if not raw_items or len(raw_items) != items.Count():
                raise ValueError("missing build geometry")
            for raw in raw_items:
                items.MoveNext()
                visit(items.GetCurrent().GetObjectResource(), transform(raw.get("transform")), set())
            size = maximum - minimum
            if not np.isfinite(size).all() or np.max(size) <= 0:
                raise ValueError("empty geometry")
        except Exception as error:
            raise HTTPException(422, "无法读取有效的 3MF 模型，或模型结构超出支持范围。请导出单个装配的 3MF/STL；不能上传仅含切片指令的文件。") from error
        return assert_dimensions(size)


def can_queue_for_manual_3mf_review(path, error):
    """复杂但结构还算可信的文件可以留给人工看。"""
    cause = error.__cause__
    reason = str(cause or error)
    if not any(text in reason for text in REVIEWABLE_PARSE_REASONS):
        return False
    try:
        with ZipFile(path) as package:
            entries = package.infolist()
            if not entries or len(entries) > MAX_ENTRIES:
                return False
            seen, has_reviewable_model = set(), False
            for item in entries:
                if item.filename in seen or unsafe_zip_entry(item):
                    return False
                seen.add(item.filename)
                if not item.filename.lower().endswith(".model"):
                    continue
                with package.open(item) as source:
                    if item.file_size <= MAX_XML:
                        content = source.read(item.file_size + 1)
                        if len(content) != item.file_size or b"<!DOCTYPE" in content.upper() or b"<!ENTITY" in content.upper() or b"\x00" in content:
                            return False
                        root = ET.fromstring(content.decode("utf-8-sig"))
                        has_build_item = any(local_name(node.tag) == "item" for node in root.iter())
                        has_reviewable_model = local_name(root.tag) == "model" and root.get("unit", "millimeter") in UNITS and has_build_item
                    else:
                        head = source.read(65536)
                        has_reviewable_model = b"\x00" not in head and b"<" in head and b"model" in head.lower()
                if has_reviewable_model:
                    return True
    except (BadZipFile, ET.ParseError, OSError, UnicodeError, ValueError):
        return False
    return False

File: frontend/src/utils/threeMF.js
import { Unzip, UnzipInflate } from 'fflate'
import { XMLParser, XMLValidator } from 'fast-xml-parser'
import { Matrix4, Vector3 } from 'three'

// 这些上限和后端保持一致，主要是防压缩包把内存撑满。
const MAX_EXPANDED = 128 * 1024 * 1024
const MAX_XML = 32 * 1024 * 1024
const MAX_ENTRIES = 2048
const MAX_INSTANCES = 4096
const MAX_TRIANGLES = 500000
const MAX_VERTICES = 1500000
const MAX_PREVIEW_TRIANGLES = 250000
const units = { micron: 0.001, millimeter: 1, centimeter: 10, inch: 25.4, foot: 304.8, meter: 1000 }
const message = '无法读取有效的 3MF 模型，或模型结构超出支持范围。请导出单个装配的 3MF/STL；不能上传仅含切片指令的文件。'
const reviewableReasons = [
  'expanded archive too large',
  'XML too large',
  'expanded mesh too large',
  'assembly too complex',
  'beam lattice geometry unsupported',
  'mixed part units',
  'mesh geometry required',
  'incomplete components',
]

function canQueueForManualReview(cause) {
  const reason = String(cause?.message || cause || '')
  return reviewableReasons.some((text) => reason.includes(text))
}

function readParts(bytes) {
  const parts = new Map()
  const names = new Set()
  let expanded = 0, xmlSize = 0, finished = 0
  const parser = new XMLParser({ ignoreAttributes: false, removeNSPrefix: true, parseTagValue: false, maxNestedTags: 100,
    parseAttributeValue: false, processEntities: false,
    isArray: (name) => ['object', 'vertex', 'triangle', 'component', 'item', 'Relationship'].includes(name) })
  const unzip = new Unzip((file) => {
    const name = file.name
    if (names.has(name) || names.size >= MAX_ENTRIES || name.startsWith('/') || /[\\:]/.test(name) || name.split('/').includes('..')) throw new Error('unsafe package')
    names.add(name)
    const xml = /\.(model|xml|rels)$/i.test(name)
    const chunks = []
    let size = 0
    file.ondata = (error, data, final) => {
      if (error) throw error
      size += data.length
      expanded += data.length
      if (xml) xmlSize += data.length
      if (expanded > MAX_EXPANDED || xmlSize > MAX_XML) throw new Error('expanded archive too large')
      if (xml) chunks.push(data)
      if (!final) return
      finished++
      if (file.originalSize !== undefined && size !== file.originalSize) throw new Error('incomplete archive')
      if (!xml) return
      const content = new Uint8Array(size)
      let offset = 0
      for (const chunk of chunks) { content.set(chunk, offset); offset += chunk.length }
      const text = new TextDecoder('utf-8', { fatal: true }).decode(content)
      if (/<!DOCTYPE|<!ENTITY|\u0000/i.test(text) || XMLValidator.validate(text) !== true) throw new Error('unsafe XML')
      const document = parser.parse(text)
      if (/\.rels$/i.test(name) && document.Relationships?.Relationship?.some((r) => r['@_TargetMode']?.toLowerCase() === 'external')) throw new Error('external relationship')
      if (/\.model$/i.test(name)) {
        if (!document.model || !Object.hasOwn(units, document.model['@_unit'] || 'millimeter')) throw new Error('invalid model')
        const nodes = [document.model]
        while (nodes.length) {
          const node = nodes.pop()
          for (const [tag, value] of Object.entries(node)) {
            if (tag.toLowerCase() === 'beamlattice') throw new Error('beam lattice geometry unsupported')
            if (value && typeof value === 'object') nodes.push(value)
          }
        }
        parts.set('/' + name, document.model)
      }
    }
    file.start()
  })
  unzip.register(UnzipInflate)
  // 分块解压，超限时可以早点停。
  for (let start = 0; start < bytes.length; start += 16384) unzip.push(bytes.subarray(start, start + 16384), start + 16384 >= bytes.length)
  if (finished !== names.size || !parts.size) throw new Error('incomplete archive')
  return parts
}

function transform(text) {
  if (text === undefined) return new Matrix4()
  const values = text.trim().split(/\s+/).map(Number)
  if (values.length !== 12 || !values.every(Number.isFinite)) throw new Error('invalid transform')
  // 3MF 和 Three.js 的矩阵方向不一样，这里转一下。
  const v = values
  return new Matrix4().set(v[0], v[3], v[6], v[9], v[1], v[4], v[7], v[10], v[2], v[5], v[8], v[11], 0, 0, 0, 1)
}

export async function read3MF(buffer, { preview = true } = {}) {
  const handles = []
  const keep = (handle) => { handles.push(handle); return handle }
  let sdk
  try {
    const bytes = new Uint8Array(buffer)
    const parts = readParts(bytes)
    // 只有读 3MF 时才加载 WASM。
    const { default: initialize } = await import('@3mfconsortium/lib3mf')
    // 虚拟程序名用纯英文，中文目录下更稳。
    sdk = await initialize({ thisProgram: '/model-inspector' })
    const wrapper = keep(new sdk.CWrapper())
    const model = keep(wrapper.CreateModel())
    const reader = keep(model.QueryReader('3mf'))
    sdk.FS.writeFile('/input.3mf', bytes)
    reader.ReadFromFile('/input.3mf')
    const root = parts.get(keep(model.RootModelPart()).GetPath())
    const unit = units[root['@_unit'] || 'millimeter']
    const rawObjects = new Map()
    for (const [part, document] of parts) {
      if (units[document['@_unit'] || 'millimeter'] !== unit) throw new Error('mixed part units')
      for (const object of document.resources?.object || []) {
        const key = part + '#' + Number(object['@_id'])
        if (rawObjects.has(key)) throw new Error('duplicate object')
        rawObjects.set(key, object)
      }
    }
    const min = [Infinity, Infinity, Infinity], max = [-Infinity, -Infinity, -Infinity]
    const meshes = new Map(), instances = []
    let visits = 0, triangles = 0, vertexCount = 0
    const point = new Vector3()
    function visit(object, matrix, ancestors) {
      const key = keep(object.PackagePart()).GetPath() + '#' + object.GetModelResourceID()
      visits += 1
      if (visits > MAX_INSTANCES || ancestors.size >= 32) throw new Error('assembly too complex')
      if (ancestors.has(key)) throw new Error('cyclic assembly unsupported')
      const raw = rawObjects.get(key)
      if (object.IsMeshObject()) {
        const mesh = keep(model.GetMeshObjectByID(object.GetUniqueResourceID()))
        triangles += mesh.GetTriangleCount()
        vertexCount += mesh.GetVertexCount()
        if (triangles > MAX_TRIANGLES || vertexCount > MAX_VERTICES) throw new Error('expanded mesh too large')
        if (!meshes.has(key)) {
          // 坐标从 XML 取原值，免得 Float32 漏掉一点点超限。
          const vertices = (raw.mesh?.vertices?.vertex || []).map((v) => ['x', 'y', 'z'].map((axis) => Number(v['@_' + axis])))
          const faces = (raw.mesh?.triangles?.triangle || []).map((t) => ['v1', 'v2', 'v3'].map((axis) => Number(t['@_' + axis])))
          if (vertices.length !== mesh.GetVertexCount() || faces.length !== mesh.GetTriangleCount() || vertices.length < 3 || !faces.length
            || vertices.some((v) => !v.every(Number.isFinite)) || faces.some((f) => !f.every((i) => Number.isInteger(i) && i >= 0 && i < vertices.length))) throw new Error('invalid mesh')
          meshes.set(key, { vertices, faces })
        }
        const geometry = meshes.get(key)
        for (const v of geometry.vertices) {
          point.fromArray(v).applyMatrix4(matrix).multiplyScalar(unit)
          for (let axis = 0; axis < 3; axis++) {
            const value = point.getComponent(axis)
            if (!Number.isFinite(value)) throw new Error('invalid transformed vertex')
            min[axis] = Math.min(min[axis], value)
            max[axis] = Math.max(max[axis], value)
          }
        }
        if (preview && triangles <= MAX_PREVIEW_TRIANGLES) instances.push({ ...geometry, matrix })
      } else if (object.IsComponentsObject()) {
        const components = keep(model.GetComponentsObjectByID(object.GetUniqueResourceID()))
        const children = raw.components?.component || []
        if (children.length !== components.GetComponentCount()) throw new Error('incomplete components')
        for (let i = 0; i < children.length; i++) {
          const child = keep(keep(components.GetComponent(i)).GetObjectResource())
          visit(child, matrix.clone().multiply(transform(children[i]['@_transform'])), new Set([...ancestors, key]))
        }
      } else throw new Error('mesh geometry required')
    }
    const build = keep(model.GetBuildItems())
    const rawItems = root.build?.item || []
    if (!rawItems.length || rawItems.length !== Number(build.Count())) throw new Error('missing build geometry')
    for (const item of rawItems) {
      build.MoveNext()
      visit(keep(keep(build.GetCurrent()).GetObjectResource()), transform(item['@_transform']), new Set())
    }
    const dimensions = max.map((value, axis) => value - min[axis])
    if (!dimensions.every(Number.isFinite) || Math.max(...dimensions) <= 0) throw new Error('empty geometry')
    let positions = null
    if (preview && triangles <= MAX_PREVIEW_TRIANGLES) {
      positions = new Float32Array(triangles * 9)
      let index = 0
      for (const instance of instances) for (const face of instance.faces) for (const vertex of face) {
        point.fromArray(instance.vertices[vertex]).applyMatrix4(instance.matrix).multiplyScalar(unit)
        // 先移回原点，远处的小模型预览不会糊掉。
        for (let axis = 0; axis < 3; axis++) positions[index++] = point.getComponent(axis) - min[axis]
      }
    }
    return { dimensions, positions, triangles }
  } catch (cause) {
    const error = new Error(message, { cause })
    error.reviewRequired = canQueueForManualReview(cause)
    throw error
  } finally {
    for (const handle of handles.reverse()) handle.delete()
    if (sdk?.FS.analyzePath('/input.3mf').exists) sdk.FS.unlink('/input.3mf')
  }
}
